src/bucket/complete_multipart_upload.gleam

```diff
@@ -0,0 +1,101 @@
+import bucket.{type BucketError, type Credentials}
+import bucket/internal
+import bucket/internal/xml
+import gleam/bit_array
+import gleam/http
+import gleam/http/request.{type Request, Request}
+import gleam/http/response.{type Response}
+import gleam/int
+import gleam/list
+import gleam/option
+import gleam/result
+import gleam/string_builder
+import xmb
+
+pub type CompleteMultipartUploadResult {
+  CompleteMultipartUploadResult(
+    location: String,
+    bucket: String,
+    key: String,
+    etag: String,
+  )
+}
+
+/// The parameters for the API request
+pub type RequestBuilder {
+  RequestBuilder(
+    bucket: String,
+    key: String,
+    upload_id: String,
+    parts: List(Part),
+  )
+}
+
+pub type Part {
+  Part(part_number: Int, etag: String)
+}
+
+pub fn request(
+  bucket bucket: String,
+  key key: String,
+  upload_id upload_id: String,
+  parts parts: List(Part),
+) -> RequestBuilder {
+  RequestBuilder(bucket:, key:, upload_id:, parts:)
+}
+
+pub fn build(builder: RequestBuilder, creds: Credentials) -> Request(BitArray) {
+  let body =
+    xmb.x(
+      "CompleteMultipartUpload",
+      [#("xmlns", "http://s3.amazonaws.com/doc/2006-03-01/")],
+      list.map(builder.parts, fn(part) {
```

Comment:
What's the relationship between the part number and the ordering of the XML elements?

Reply:
They must be ordered, otherwise we get this error
```
{error,
    {s3_error,400,
        {error_object,<<"InvalidPartOrder">>,
            <<"The list of parts was not in ascending order. The parts list must be specified in order by part number.">>,
            <<"/multipart1/test/myfile">>,<<"17FC9D83BD0130A6">>}}},
```

Reply:
Why do we ask the programmer to give a number at all if they always have to count up from 1? Couldn't we accept just the etags and add then number ourselves? That way it's not longer possible for the programmer to make a mistake by providing an invalid state.

What am I missing here?

Reply:
I've since discovered that S3 needs the part number to recognize the part, ETag is not enough.
The part numbers may not start at 1 and may not always increment by 1: It's possible to drop some parts in CompleteMultipartUpload, but they must have correct PartNumbers and they must be in ascending order.

I added a test for this in df9c3b3